driver_test.go

```diff
@@ -12,26 +12,220 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package spannerdriver_test
+package spannerdriver
 
 import (
+	"cloud.google.com/go/spanner"
+	"context"
 	"database/sql"
+	"fmt"
 	"log"
+	"os"
+	"reflect"
+	"runtime/debug"
+	"testing"
 
-	spannerdriver "github.com/rakyll/go-sql-driver-spanner"
+	// api/lib packages not imported by driver
+	adminapi "cloud.google.com/go/spanner/admin/database/apiv1"
 	"google.golang.org/api/option"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+	"google.golang.org/grpc"
 )
 
-func ExampleDriver() {
-	driver := &spannerdriver.Driver{
-		Options: []option.ClientOption{
-			option.WithCredentialsFile("/path/to/service-account-key.json"),
-		},
+var (
+	project  string
+	instance string
+	dbname   string
+	dsn      string
+	projset  bool
+	instset  bool
+	dbset    bool
+)
+
+type Connector struct {
+	ctx         context.Context
+	client      *spanner.Client
+	adminClient *adminapi.DatabaseAdminClient
+}
+
+func NewConnector() (*Connector, error) {
+
+	ctx := context.Background()
+
+	adminClient, err := adminapi.NewDatabaseAdminClient(
+		ctx,
+		option.WithoutAuthentication(),
+		option.WithEndpoint("0.0.0.0:9010"),
+		option.WithGRPCDialOption(grpc.WithInsecure()))
+	if err != nil {
+		return nil, err
+	}
+
+	dataClient, err := spanner.NewClient(ctx, dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	curs := &Connector{
+		ctx:         ctx,
+		client:      dataClient,
+		adminClient: adminClient,
+	}
+	return curs, nil
+}
+
+func (c *Connector) Close() {
+	c.client.Close()
+	c.adminClient.Close()
+}
+
+// structs for row data
+type testaRow struct {
+	A string
+	B string
+	C string
+}
+type typeTestaRow struct {
+	stringt string
+	intt    int
+	floatt  float64
+	boolt   bool
+}
+
+func init() {
+
+	// get environment variables or set to default
+	if instance, instset = os.LookupEnv("SPANNER_TEST_INSTANCE"); !instset {
+		instance = "test-instance"
+	}
+	if project, projset = os.LookupEnv("SPANNER_TEST_PROJECT"); !projset {
+		project = "test-project"
+	}
+	if dbname, dbset = os.LookupEnv("SPANNER_TEST_DBNAME"); !dbset {
+		dbname = "gotest"
+	}
+
+	// derive data source name
+	dsn = "projects/" + project + "/instances/" + instance + "/databases/" + dbname
+}
+
+// Executes DDL statements
+func executeDdlApi(curs *Connector, ddls []string) {
```

Comment:
nit: Maybe we could have different functions for `createDatabase` and `updateDatabaseDdl`?

Reply:
I'll revisit when/if I can get DDL to work from the driver with the emulator 🙂